registries/_format/byte.md

```diff
@@ -12,7 +12,7 @@ layout: default
 
 Base type: `{{ page.base_type }}`.
 
-The `{{page.slug}}` format represents any sequece of octets encoded as a base64 string. This format entry is to ensure future versions of OpenAPI maintain compatibility with [OpenAPI 3.0.x](https://spec.openapis.org/oas/v3.0.0).
+The `{{page.slug}}` format represents any sequence of octets encoded as a base64 string. This format entry is to ensure future versions of OpenAPI maintain compatibility with [OpenAPI 3.0.x](https://spec.openapis.org/oas/v3.0.0).
```

Comment:
Similarly, in OAS 3.1 this is better written as `"contentEncoding": "base64"` (which is distinct from `"base64url"`).

Reply:
I think there was a confusion for this one. We (in Kiota) treat byte at uint8. But again maybe the confusion is on our side?

Reply:
@baywet it's not clear to me how `any sequence of octets encoded as a base64 string.` is treated as `uint8`.  Do you mean you interpret the _decoded_ string as an array of `uint8`?  I don't think there are any particular constraints or guidelines on what to do with the decoded string.  This just says that it's encoded.

Reply:
sorry, I got confused with sbyte for a second by reading the path of the comment too fast. Yes kiota is mapping that to a binary format as well.
would it be fair to say that base64url is just base64 + uriEncode ?

Reply:
These things are defined by RFC 4648:

* [base64](https://www.rfc-editor.org/rfc/rfc4648#section-4)
* [base64url](https://www.rfc-editor.org/rfc/rfc4648#section-5)

I don't know what you mean by "mapping to a binary format."  I assume you are using "format" to mean something other than the `format` keyword which is the source of the format?

It might help if you explained more about your environment as you seem to be making a lot of assumptions of knowledge.  I haven't worked in a strongly typed language in over a decade, for example, and I'm not familiar with kiota.  So I don't know what you need to do with these things or what constraints you're working under.

Reply:
Thanks for pointing to the RFC, I should read those more as background learning, that just resolved a lot of confusion from my early programing days 😅.

You are correct, I've been making a lot of assumptions in my past comments about knowledge of what the client side code is supposed to look like.
Any property that's not a scalar value or a complex object is reflected on the client side as a stream or binary array, and kiota handles the "(de)serialization" automatically for the client application so the developer writing the client application doesn't need to care for/understand these concerns.

If the response content type is something "unstructured" (e.g. image) or serialized in a format we don't understand (let's say XML since we don't have serialization providers for XML yet), we project the whole response body as a stream/byte array.

Same logic applies for request bodies.

With that context in mind, I'm asserting the following:
People should avoid using base64/base64url for the whole response body and use application/octet-stream when designing APIs instead since base64 encoding will "inflate" the response payload with no added value.
However, it's fine to use that for a single property in a structured payload to avoid using multipart or requiring the client to make multiple requests.
If they end up using base64/base64url for the whole payload for whatever reason, and with 3.1 they shuold use the contentEncoding property instead of the format one.

Does that make sense?

Reply:
@baywet I will get back to you on this, I'm having to research things because we're getting into one of the most mind-numbingly complex parts of the spec.  I'd complain about the person who wrote the OAS 3.1 section in question, but that was me 😅 

I think you're more-or-less right, but there are cases not covered and I want to make sure I say the right things about when and why `format: binary` (and the equivalent construct in OAS 3.1 / draft 2020-12) can be used.

Reply:
OK, here's my thoughts on all of this:

* raw binary data is only valid as a complete payload (e.g. `Content-Type: image/png`) or as one complete part of a `multipart/form-data` payload (or perhaps other `multipart` media types, but those are not clearly addressed by OAS, so you probably have to compare them to [`multipart/form-data`](https://www.rfc-editor.org/rfc/rfc7578#section-4) and reverse engineer how to handle it).
* base64-encoded data (or other encodings) _can_ be used in those situations, but as you note, it's not a good idea as it likely inflates the payload size; these encodings were designed for MIME and 7-bit character set restrictions
* base64-encoded data is a good choice if you are embedding binary data _inside of an `application/json` payload, such as in an object property or array item_.  (You could be pathological and make an `application/json` payload that consists of a JSON string with binary encoded data, but... don't)
* base64url-encoded data is a good choice if you are embedding (presumably small) binary data _inside of a URL such as in a query string parameter_.

### Raw binary payloads

```YAML
# OAS 3.0
content:
  image/png:
    schema:
      type: string
      format: binary
```
For OAS 3.1, the simplest thing is to just not provide a schema:
```
# OAS 3.1
content:
  image/png: {}
```
However, in draft 2020-12, we allow data outside of the JSON Schema data model to be described by a JSON Schema, even though no keywords are relevant.  You just use an empty schema `{}` or its equivalent boolean schema `true`:
```
# OAS 3.1
content:
  image/png:
    schema: true
```

### Encoded binary payloads

As you noted, these don't make much sense.  However, if you wanted to use them, here is how you would do it.  Except... OAS 3.0 gives different formats in [§4.4 Data Types](https://spec.openapis.org/oas/v3.0.3#data-types) (`format: byte`) vs in the example in [§4.7.14.3 Considerations for File Uploads](https://spec.openapis.org/oas/v3.0.3#special-considerations-for-multipart-content) (`format: base64`).  Your guess is as good as mine as to what to document- maybe both?

```YAML
# OAS 3.0
content:
  image/png:
    schema:
      type: string
      format: byte    # according to §4.4
```
For OAS 3.1, this is the minimal approach (you could use other encodings than base64 if you wanted):
```YAML
# OAS 3.1
content:
  image/png:
    schema:
      type: string
      contentEncoding: base64
```
If you want a JSON Schema application to be aware of the media type of the encoded content, you need to duplicate that inside the schema with `contentMediaType`:
```YAML
# OAS 3.1
content:
  image/png:
    schema:
      type: string
      contentMediaType: image/png
      contentEncoding: base64
```

### Raw binary parts in `multipart/form-data`

In OAS 3.0 you have to use `format: binary` with an Encoding Object to show that the encoded data is `image/png`:

```YAML
# OAS 3.0
content:
  multipart/form-data:
    schema:
      type: object
        properties:
          anImage:
            type: string
            format: binary
    encoding:
      anImage:
        contentType: image/png
```
In OAS 3.1, for binary data, the easiest option is to use an Encoding Object and an empty (or `true`) schema.
```YAML
# OAS 3.1
content:
  multipart/form-data:
    schema:
      type: object
        properties:
          anImage: {}
    encoding:
      anImage:
        contentType: image/png
```
You generally would not pass this to JSON Schema implementations as they don't understand `multipart/form-data`, particularly not with binary parts, so you should avoid `contentEncoding` and `contentMediaType` (for complex reasons, unless you _also_ use an Encoding Object to set `contentType: image/png`, your multipart section will end up as `text/plain`).

### Encoded binary parts in `multipart/form-data`

In OAS 3.0 encoding is just like binary but with a different `format`

```YAML
# OAS 3.0
content:
  multipart/form-data:
    schema:
      type: object
        properties:
          anImage:
            type: string
            format: byte  # per §4.4
    encoding:
      anImage:
        contentType: image/png
```
In OAS 3.1, for you have to use both the JSON Schema keywords *and* the Encoding Object to keep the multipart part from being set to `text/plain`:
```YAML
# OAS 3.1
content:
  multipart/form-data:
    schema:
      type: object
        properties:
          anImage:
            type: string
            contentMediaType: image/png
            contentEncoding: base64
    encoding:
      anImage:
        contentType: image/png
```
You generally would not pass this to JSON Schema implementations as they don't understand `multipart/form-data`, so you could perhaps get away without `contentMediaType`, idk.

### Embedding binary payloads in JSON the WRONG way

You can't use `format: binary` to embed binary data inside of a JSON payload, because the JSON data model doesn't allow it.  So this is **wrong**:

```YAML
# OAS 3.0
content:
  application/json:
    schema:
      type: object
      properties:
        anImage:
          type: string
          format: binary     # DO NOT DO THIS!!!!!!
```
or this
```YAML
# OAS 3.1
content:
  application/json:
    schema:
      type: object
      properties:
        anImage:
          type: string
          contentMediaType: image/png
          contentEncoding: binary     # DO NOT DO THIS!
```
JSON cannot embed binary data, and this is what I was thinking of when I was getting particularly confused.  It's probably worth noting this in the `format: binary` entry.

## Embedding encoded binary data in a JSON payload (the right way)

However, if you really do want to embed a PNG in a JSON object field, you want to encode it, and `base64` is probably what you want.  I'm going to use `format: byte` per §4.4 in this example, but remember that `format: base64` is used in an OAS 3.0 example, so idk what is better-supported out there).

```YAML
# OAS 3.0
content:
  application/json:
    schema:
      type: object
      properties:
        anImage:
          type: string
          format: byte   # correct per §4.4 Data Types
```

Note that there is no way to tell from the schema or the `content` entry that there is an `image/png` in the string.  You just have to assume `application/octet-stream` or rely on human-readable documentation.

In OAS 3.1, this is fixed with `contentMediaType`, which goes alongside `contentEncoding`:

```YAML
# OAS 3.1
content:
  application/json:
    schema:
      type: object
      properties:
        foo:
          type: string
          contentMediaType: image/png
          contentEncoding: base64
```
### Embedding in a URL query parameter

If you are embedding a (presumably very small) image in a URL query parameter, then you might want to use `base64url` so you don't end up with multiple layers of encoding going on.  This is only possible in OAS 3.1; in OAS 3.0 your only option is `base64`.

In OAS 3.0 you also need to use `content` instead of `schema` in order to record the encoded media type:

```YAML
parameters:
  - name: anImage
    in: query
    content:
      image/png:
        schema:
          type: string
          format: byte    # per §4.4
```

In OAS 3.1, you can just rely on the JSON Schema fields, although OAS implementations might prefer that you use `content`.

```YAML
# OAS 3.1, JSON Schema only
parameters:
  - name: anImage
    in: query
    schema:
      type: string
      contentMediaType: image/png
      contentEncoding: base64url
```
```YAML
# OAS 3.1, content object and JSON Schema
parameters:
  - name: anImage
    in: query
    content:
      image/png:
        schema:
          type: string
          contentMediaType: image/png
          contentEncoding: base64url
```
```YAML
# OAS 3.1, content object only
parameters:
  - name: anImage
    in: query
    content:
      image/png: {}
```

Reply:
@handrews please could you publish that as a blog post? It deserves a wider audience. We could shepherd it through as an openapis.org post.

Reply:
this is gold! Thank you so much. I'll try to capture it properly in this PR.
Agreed with @MikeRalphson this could serve as a blog post, and we could leverage that blog post to announce that formats are documented once this PR is merged, what do you think?

Reply:
I've just pushed the changes, that I hope capture most of this information. I did have a couple of questions to make sure I didn't miss anything.

## query parameters

I believe the OAS 3.0 example was meant to use base64url since byte/base64 would not be URL safe?
Would the last OAS3.1 example work since there is no encoding?

## multi-part form
In the OAS 3.1 example, could people use contentMediaType instead?

Reply:
@MikeRalphson @baywet yes, I will definitely write a blog post.  I actually started writing this as one yesterday and then realized there are a few more things not relevant to the PR that would need to go in it, and that I didn't have time to come up with the appropriate intro / conclusion prose.  But I'll work on that, probably next week.

@MikeRalphson I'll follow up with you on the OpenAPI slack about where/how to publish it.

@baywet regarding your quetions:

* OAS 3.0 does not define a `base64url` format.  Of course, you could add it if you wanted to as it comes from that RFC :-)

* That last OAS 3.1 example might not work in tools that expect the Encoding Object, so yeah I wouldn't recommend it.  I might leave it out of the blog post (that's part of what I need to think through- how much of the different possible ways to write things to include).

* For reasons that I think had to do with a.) compatibility with OAS 3.0 and b.) the fact that this kind of got squeezed in at the last minute, if the schema says `"type": "string"` then unless you also use an Encoding Object, OpenAPI specifies the content type to be `text/plain`.  I don't recall why we didn't allow `contentMediaType` to override that, but I think I was just trying to make minimal changes.  I was also kind of stepping back from work on OAS and JSON Schema around this time (fall 2020) so I might have meant to follow up on it and then just wasn't able to (because... 2020).

Reply:
ok thanks. I believe we're on the same page. I'll resolve this long convo, don't hesitate to open new comments on defects on pages if needed.